core/profiles/src/main/java/software/amazon/awssdk/profiles/ProfileFileSupplier.java

```diff
@@ -137,7 +138,7 @@ public ProfileFile get() {
                 return  currentAggregateProfileFile.get();
             }
 
-            private boolean didSuppliedValueChange(Supplier<ProfileFile> supplier) {
+            private synchronized boolean didSuppliedValueChange(Supplier<ProfileFile> supplier) {
```

Comment:
Should we just wrap currentValuesBySupplier in a Collections.synchronizedMap instead of synchronizing ourselves?

If there's some problem with that:

Can we synchronize just around the writes to the currentValuesBySupplier so that we don't hold the lock while calling the supplier?

```java
ProfileFile current;
synchronized (currentValuesBySupplier) {
    current = currentValuesBySupplier.put(supplier, next);
}
```

I think we also need to synchronize our iterating over the values on line 150, since modification of the collection during iteration is undefined.

```java
synchronized (currentValuesBySupplier) {
    currentValuesBySupplier.values().forEach(aggregator::addFile);
}
```

Reply:
I agree the easiest solution is just to wrap into Collections.synchronizedMap.

Yep, I missed the synchronized keyword on line 150.